chanfitness/chaneventstore.go

```diff
@@ -0,0 +1,308 @@
+		// tracking events for it anyway.
+		return
+	}
+
+	eventLog.closedAt = eventLog.now()
+}
+
+// peerEvent adds a peer online or offline event to all channels we currently
+// have open with a peer.
+func (c *ChannelEventStore) peerEvent(peer [33]byte, event eventType) {
+	c.Lock()
+	defer c.Unlock()
+
+	// Track current online status of peers in the channelEventStore.
+	peerIsOnline := true
+	if event == peerOfflineEvent {
+		peerIsOnline = false
+	}
+	c.peers[peer] = peerIsOnline
+
+	for _, eventLog := range c.channels {
+		if eventLog.peer == peer {
+			eventLog.add(event)
+		}
+	}
+}
+
+// monitorChannelEvents consumes the events channel and updates the event store
+// with channel opened/closed and peer online/offline events.
+func (c *ChannelEventStore) monitorChannelEvents(events <-chan interface{}) {
+	defer c.wg.Done()
+
+	// Consume events until the channel is closed.
+	for e := range events {
+		switch event := e.(type) {
+		// A new channel has been opened, we must add the channel to the
+		// store and record a channel open event.
+		case channelnotifier.OpenChannelEvent:
+			chanID := event.Channel.ShortChanID().ToUint64()
+
+			var peerKey [33]byte
+			copy(
+				peerKey[:],
+				event.Channel.IdentityPub.SerializeCompressed(),
+			)
+
+			c.addChannel(chanID, peerKey)
+
+		// A channel has been closed, we must remove the channel from the
+		// store and record a channel closed event.
+		case channelnotifier.ClosedChannelEvent:
+			c.closeChannel(event.CloseSummary.ShortChanID.ToUint64())
+
+		// We have reestablished a connection with our peer, and should
+		// record an online event for any channels with that peer.
+		case peernotifier.PeerOnlineEvent:
+			c.peerEvent(event.PubKey, peerOnlineEvent)
+
+		// We have lost a connection with our peer, and should record an
+		// offline event for any channels with that peer.
+		case peernotifier.PeerOfflineEvent:
+			c.peerEvent(event.PubKey, peerOfflineEvent)
+		}
+	}
+}
+
+// GetLifespan returns the opening and closing time observed for a channel and
+// a boolean to indicate whether the channel is known the the event store. If the
+// channel is still open, a zero close time is returned.
+func (c *ChannelEventStore) GetLifespan(chanID uint64) (time.Time, time.Time, bool) {
+	c.Lock()
+	defer c.Unlock()
+
+	channel, ok := c.channels[chanID]
+	if !ok {
+		return time.Time{}, time.Time{}, false
+	}
+
+	return channel.openedAt, channel.closedAt, true
+}
+
+// GetUptime returns the uptime of a channel over a period and a boolean to
+// indicate whether it was found in the channel event store. If a zero start or
+// end time is given, the function returns zero and false to indicate that
+// uptime was not calculated.
+func (c *ChannelEventStore) GetUptime(chanID uint64, startTime,
+	endTime time.Time) (time.Duration, bool) {
+
+	// Return early if the period specified has a zero start or end time.
+	if startTime.IsZero() || endTime.IsZero() {
```

Comment:
shouldn't `startTime` be allowed set to zero to indicate start as early as possible?

Reply:
It should! 

I'm still going to err on a 0 `endtime` though? It's easy enough to pass `now()` in for endtime.

Reply:
Do we even need to support custom `endtime` (and `startime`?)? Would it simplify things if `Uptime` just returns the uptime between `channel open` and `min(channelclose, now)` ?

Reply:
I think that it may be nice to have for longer running channels, because we may care more about recent uptime periods?

Could update this func to not take params so it's simpler for callers (which I like) but leave the functionality there in case we want it?

Reply:
Ok good point, can keep `startTime` and `endtime` options I think :)